packages/crypto/tests/wycheproof_secp256k1.rs

```diff
@@ -0,0 +1,412 @@
+use serde::Deserialize;
+
+// See ./testdata/wycheproof/README.md for how to get/update those files
+const SECP256K1_SHA256: &str = "./testdata/wycheproof/ecdsa_secp256k1_sha256_test.json";
+const SECP256K1_SHA512: &str = "./testdata/wycheproof/ecdsa_secp256k1_sha512_test.json";
+const SECP256K1_SHA3_256: &str = "./testdata/wycheproof/ecdsa_secp256k1_sha3_256_test.json";
+const SECP256K1_SHA3_512: &str = "./testdata/wycheproof/ecdsa_secp256k1_sha3_512_test.json";
+
+#[derive(Deserialize, Debug)]
+#[serde(rename_all = "camelCase")]
+struct File {
+    number_of_tests: usize,
+    test_groups: Vec<TestGroup>,
+}
+
+#[derive(Deserialize, Debug)]
+#[serde(rename_all = "camelCase")]
+struct TestGroup {
+    public_key: Key,
+    tests: Vec<TestCase>,
+}
+
+#[derive(Deserialize, Debug)]
+#[serde(rename_all = "camelCase")]
+struct Key {
+    uncompressed: String,
+}
+
+#[derive(Deserialize, Debug)]
+#[serde(rename_all = "camelCase")]
+struct TestCase {
+    tc_id: u32,
+    comment: String,
+    msg: String,
+    sig: String,
+    // "acceptable", "valid" or "invalid"
+    result: String,
+}
+
+fn read_file(path: &str) -> File {
+    use std::fs::File;
+    use std::io::BufReader;
+
+    // Open the file in read-only mode with buffer.
+    let file = File::open(path).unwrap();
+    let reader = BufReader::new(file);
+
+    serde_json::from_reader(reader).unwrap()
+}
+
+mod hashers {
+    use sha2::{Digest, Sha256, Sha512};
+    use sha3::{Sha3_256, Sha3_512};
+
+    pub fn sha256(data: &[u8]) -> [u8; 32] {
+        Sha256::digest(data).into()
+    }
+
+    // ecdsa_secp256k1_sha512 requires truncating to 32 bytes
+    pub fn sha512(data: &[u8]) -> [u8; 32] {
+        let hash = Sha512::digest(data).to_vec();
+        hash[..32].try_into().unwrap()
+    }
+
+    pub fn sha3_256(data: &[u8]) -> [u8; 32] {
+        Sha3_256::digest(data).into()
+    }
+
+    // ecdsa_secp256k1_sha3_512 requires truncating to 32 bytes
+    pub fn sha3_512(data: &[u8]) -> [u8; 32] {
+        let hash = Sha3_512::digest(data).to_vec();
+        hash[..32].try_into().unwrap()
+    }
+}
+
+#[test]
+fn ecdsa_secp256k1_sha256() {
+    let mut tested: usize = 0;
+    let File {
+        number_of_tests,
+        test_groups,
+    } = read_file(SECP256K1_SHA256);
+    assert!(number_of_tests >= 463, "Got unexpected number of tests");
+
+    for group in test_groups {
+        let public_key = hex::decode(group.public_key.uncompressed).unwrap();
+        assert_eq!(public_key.len(), 65);
+
+        for tc in group.tests {
+            tested += 1;
+            assert_eq!(tc.tc_id as usize, tested);
+            // eprintln!("Test case ID: {}", tc.tc_id);
+
+            match tc.result.as_str() {
+                "valid" | "acceptable" => {
+                    let message = hex::decode(tc.msg).unwrap();
+                    let message_hash = hashers::sha256(&message);
+                    let der_signature = hex::decode(tc.sig).unwrap();
```

Comment:
This code is the same in both match arms, so we can just move it up, out of the match. Same for the other tests.

Reply:
Oh, yeah, nice. Done